### PR TITLE
Use BCrypt by default as encoder #1312

### PR DESCRIPTION
**Pull request by e-moe**
Based on work of @mvanbaak (https://github.com/silexphp/Silex/pull/1316) with applied comments from @mathroc and fixed tests.

### CODE REVIEW

**Comment by e-moe** [comment]
Merge conflict fixed.

File: doc/providers/security.rst [hunk 1]
```diff
@@ -552,18 +558,16 @@ sample users::
...
 
-    use Symfony\Component\Security\Core\Encoder\MessageDigestPasswordEncoder;
+    use Symfony\Component\Security\Core\Encoder\BCryptPasswordEncoder;
 
     $app['security.encoder.digest'] = function ($app) {
-        // use the sha1 algorithm
-        // don't base64 encode the password
-        // use only 1 iteration
-        return new MessageDigestPasswordEncoder('sha1', false, 1);
+        // use only 10 iterations
```
Comment (mathroc) [inline doc/providers/security.rst]: with bcrypt, the number of iterations is actually 2 to the power of `$cost`

**Comment by mathroc** [comment]
looks good to me! :+1:

File: doc/providers/security.rst (hunk 1 above)
Comment (e-moe) [inline doc/providers/security.rst]: thanks for comment @mathroc, updated

**Comment by mvanbaak** [comment]
Thank you for doing my work. If only time did allow me to put my actions where my mouth is :(
(+1)

File: doc/providers/security.rst [hunk 2]
```diff
@@ -552,18 +558,16 @@ sample users::
...
-service::
+By default, Silex uses the ``BCrypt`` algorithm to encode passwords.
+Additionally, the password is encoded multiple times.
+You can change these defaults by overriding the encoding cost
+``security.encoder.bcrypt.cost`` or ``security.default_encoder`` service::
 
-    use Symfony\Component\Security\Core\Encoder\MessageDigestPasswordEncoder;
+    use Symfony\Component\Security\Core\Encoder\BCryptPasswordEncoder;
 
     $app['security.encoder.digest'] = function ($app) {
```
Comment (fabpot) [inline doc/providers/security.rst]: should be `bcrypt` here, right?

**Comment by fabpot** [comment]
Would it makes sense to instead use `Pbkdf2PasswordEncoder`? Or at least add a sensible configuration for it by default in the provider?

File: doc/providers/security.rst (hunk 2 above)
Comment (mathroc) [inline doc/providers/security.rst]: indeed, `security.encoder.bcrypt` or maybe `security.default_encoder` (as it could be replaced with anything)

**Comment by mathroc** [comment]
@fabpot I don't have enough knowledge to argue for bcrypt or pbkdf2 but [the php documentation on hash_pbkdf2](php.net/hash_pbkdf2) says : 
>  The PBKDF2 method can be used for hashing passwords for storage. However, it should be noted that password_hash() or crypt() with CRYPT_BLOWFISH are better suited for password storage. 

But yes, a `security.encoder.pbkdf2` could be added as it's also widely used

@e-moe what do you think ?

**Comment by e-moe** [comment]
@mathroc I'm not a security expert either. But I would say let's give users possibility to choose. I will update my pull request

**Comment by e-moe** [comment]
updated to meet comments from @fabpot and @mathroc

File: doc/providers/security.rst [hunk 3]
```diff
@@ -10,6 +10,8 @@ Parameters
 * **security.hide_user_not_found** (optional): Defines whether to hide user not
   found exception or not. Defaults to ``true``.
 
+* **security.encoder.bcrypt.cost** (optional): Defines BCrypt password encoder cost. Defaults to 15.
```
Comment (stof) [inline doc/providers/security.rst]: Symfony's SecurityBundle defaults to 13. Shouldn't we be consistent here ?

Comment (e-moe) [inline doc/providers/security.rst]: updated

**Comment by fabpot** [comment]
Thank you @e-moe.

**Pull request merged**